### PR TITLE
[JENKINS-31512] Added support for XShell plugin

### PR DESCRIPTION
**Pull request by v1v**
Added support for [XShell](https://wiki.jenkins-ci.org/display/JENKINS/XShell+Plugin) plugin: 
- https://issues.jenkins-ci.org/browse/JENKINS-31512

### CODE REVIEW

**Comment by v1v** [comment]
There are two fields which are missing from the non default configuration:
- _Regex to kill job_
- _Time allocated_

I'll add them as optional ones

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/StepContext.groovy [hunk 1]
```diff
@@ -42,6 +42,26 @@ class StepContext extends AbstractExtensibleContext {
     }
 
     /**
+     * Runs a xshell script.
+     *
+     * @since 1.41
+     */
+    @RequiresPlugin(id = 'xshell', minimumVersion = '0.10')
+    void xshell(String commandLine = null, Boolean executeFromWorkingDir = false, String regexToKill = null,
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/StepContext.groovy]: Replace the parameters by a nested context. Contexts are more convenient for users and allow to specify selected parameters.

**Comment by daspilker** [comment]
And please squash your commits when you are done.

**Comment by v1v** [comment]
Hi @daspilker ,

I've already changed to use nested context and squashed those commits

Cheers

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/step/StepContext/xshell.groovy [hunk 2]
```diff
@@ -0,0 +1,22 @@
+job('example') {
+    steps {
+        xshell {
+            commandLine ('echo Hi!')
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/step/StepContext/xshell.groovy]: Do not add a space between method name and opening bracket.

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/XShellContext.groovy [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    boolean executeFromWorkingDir = false
+    String regexToKill = null
+    int timeAllocated = 0
+
+    protected XShellContext(JobManagement jobManagement) {
+        super(jobManagement)
+    }
+
+    /**
+     * Adds xshell command.
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/XShellContext.groovy]: Improve the GroovyDoc for all methods. Have a look at the xshell plugin's help texts to get an idea.

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/XShellContext.groovy [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+     */
+    void commandLine(String commandLine) {
+        this.commandLine = commandLine
+    }
+
+    /**
+     * Execute command from Working Dir
+     */
+    void executeFromWorkingDir(boolean executeFromWorkingDir) {
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/XShellContext.groovy]: The UI description for this is "Executable is in workspace dir", so "executeFromWorkingDir" is misleading. Rename this to "executableInWorkspaceDir". And set the parameters default value to true: `void executeFromWorkingDir(boolean executeFromWorkingDir = true) { ... }`

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/step/StepContextSpec.groovy [hunk 5]
```diff
@@ -61,6 +61,41 @@ class StepContextSpec extends Specification {
         1 * jobManagement.requireMinimumPluginVersion('ssh', '1.3')
     }
 
+    def 'call xshell method with minimum options'() {
+        when:
+        context.xshell {
+            commandLine 'echo "Hello"'
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/step/StepContextSpec.groovy]: the commandLine option is not required by your implementation, so this does not test the minimum options.

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/StepContext.groovy [hunk 6]
```diff
@@ -42,6 +42,27 @@ class StepContext extends AbstractExtensibleContext {
...
+     * Runs a xshell script.
+     *
+     * @since 1.41
+     */
+    @RequiresPlugin(id = 'xshell', minimumVersion = '0.10')
+    void xshell(@DslContext(XShellContext) Closure xShellClosure) {
+        XShellContext xShellContext = new XShellContext(jobManagement)
+        ContextHelper.executeInContext(xShellClosure, xShellContext)
+        stepNodes << new NodeBuilder().'hudson.plugins.xshell.XShellBuilder' {
+            delegate.executableInWorkspaceDir(xShellContext.executableInWorkspaceDir)
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/StepContext.groovy]: The XML element is still called `executeFromWorkingDir`. Do not rename the XML element names. DSL methods names and XML element names must not match. Use sane name for the DSL methods using the UI labels as an orientation. The XML element names are just implementation details. And please test your changes before pushing.

Comment (v1v) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/step/StepContext.groovy]: Awesome! I did get confused when I did replace those matches, thanks for that spot! I've already changed, squashed and pushed
Thanks

**Pull request merged**